api/allocations_exec.go

```diff
@@ -0,0 +1,231 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type execSession struct {
+	client  *Client
+	alloc   *Allocation
+	task    string
+	tty     bool
+	command []string
+
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+
+	terminalSizeCh <-chan TerminalSize
+
+	q *QueryOptions
+}
+
+func (s *execSession) run(ctx context.Context) (exitCode int, err error) {
+	ctx, cancelFn := context.WithCancel(ctx)
+	defer cancelFn()
+
+	conn, err := s.startConnection()
+	if err != nil {
+		return -2, err
+	}
+	defer conn.Close()
+
+	sendErrCh := s.startTransmit(ctx, conn)
+	exitCh, recvErrCh := s.startReceiving(ctx, conn)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return -2, ctx.Err()
+		case exitCode := <-exitCh:
+			return exitCode, nil
+		case recvErr := <-recvErrCh:
+			// drop websocket code, not relevant to user
+			if wsErr, ok := recvErr.(*websocket.CloseError); ok && wsErr.Text != "" {
+				return -2, errors.New(wsErr.Text)
+			}
+
+			return -2, recvErr
+		case sendErr := <-sendErrCh:
+			return -2, fmt.Errorf("failed to send input: %w", sendErr)
+		}
+	}
+}
+
+func (s *execSession) startConnection() (*websocket.Conn, error) {
+	nodeClient, _ := s.client.GetNodeClientWithTimeout(s.alloc.NodeID, ClientConnTimeout, s.q)
```

Comment:
Would be helpful to have a comment explaining why `nodeClient` being nil here is acceptable - also, would it make sense to switch on the error here? E.g. we could get back `NodeDownErr` which seems like a case to bail early? 

https://github.com/hashicorp/nomad/blob/main/api/api.go#L492